internal/querycoordv2/handlers.go

```diff
@@ -147,7 +148,32 @@ func (s *Server) balanceSegments(ctx context.Context, req *querypb.LoadBalanceRe
 		}
 		tasks = append(tasks, task)
 	}
-	return task.Wait(ctx, Params.QueryCoordCfg.SegmentTaskTimeout, tasks...)
+	return waitBalanceTasks(ctx, Params.QueryCoordCfg.SegmentTaskTimeout, tasks...)
+}
+
+// waitBalanceTasks blocks until the tasks finished,
+// return the first returned err of task,
+// NOTE: this ignores the TaskStale error,
+// because the balancing segments may be compacted,
+// which should be ignored, more details #19386
+func waitBalanceTasks(ctx context.Context, timeout time.Duration, tasks ...task.Task) error {
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	var err error
+	go func() {
+		for _, t := range tasks {
+			err = t.Wait()
+			if err != nil && !errors.Is(err, task.ErrTaskStale) {
+				cancel()
+				break
+			}
+		}
+		cancel()
```

Comment:
Does tasks have order here? Looks equivalent to a plain for-loop without goroutine here

Reply:
No, the goroutine here makes the method able to return if the request's context canceled or timeout